### ISSUE TITLE
New (optional) interface in the schema package to support Swagger 2.0 generation?

### ISSUE THREAD

**Issue by smyrman** [rs/rest-layer#21]
Looking on what's missing to successfully gennerate [Swagger 2.0 specification](http://swagger.io/specification/) from a resource.Index. Most of it, likes paths, tags etc., seems trivial enough to generate from index parameters, or as a stop-gap, hardcode in the application.

However for the [definitions object](http://swagger.io/specification/#definitionsObject), I feel there is a missing piece in the schema package to do a successful auto-translation to a swagger Schema. Perhaps if there was an extra (optional) interface for Validators that was implemented by all the standard types, that could help? Perhaps something like:
```go
type Swagger interface {
    SwaggerParams() map[string]interface{}
}
```
For simple types, like bool, it probably only need to return "type", while for e.g. an integer with boundaries, it should return "maxLength", "minLength", and a "format" (if defined).

**Comment by smyrman** [comment]
On the next level, perhaps the Swagger interface could be implemented on [schema.Field](https://godoc.org/github.com/rs/rest-layer/schema#Field), which would simply run the SwaggerParams method on the validator (if present), and then supplement with it's own parameters, such as "default", "readOnly" etc. Thereafter, on schema.Schema, and then I think it's pretty much there as far as Definition Objects is concerned.

**Comment by rs** [comment]
I'm not very familiar with swagger yet, but sounds like an interesting idea. Do you think this optional interface could be useful outside of the scope of swagger? If yes, shouldn't we find a more generic name? If no, let's stick with this explicit name.

**Comment by smyrman** [comment]
JSONSchema could be a better name, as a Swagger schema is basically a subset of [JSON Schema Draft 4](http://json-schema.org/), with some very minor additions:

> This object is based on the JSON Schema Specification Draft 4 and uses a predefined subset of it. On top of this subset, there are extensions provided by this specification to allow for more complete documentation.

See http://swagger.io/specification/#schemaObject for details.

On the Schema / Field objects you could also implement https://golang.org/pkg/encoding/json/#Marshaler instead of the JSONSchema interface. Just be aware that the "required" property is a list on the Schema object rather than a boolean on the field itself.

**Comment by smyrman** [comment]
> On the Schema / Field objects you could also implement https://golang.org/pkg/encoding/json/#Marshaler instead of the JSONSchema interface. Just be aware that the "required" property is a list on the Schema object rather than a boolean on the field itself.

Again as a start, I suppose if the Validators only (as initially suggested) where to implement:
```go
type JSONSchema interface {
        func JSONSchemaParams() map[string]interface{}
}
```
Separate `swagger`  or `jsonschema` packages depending on `schema` could probably handle the remaining translation as schema.Schema and schema.Field are just structs. That way the schema package itself won't be too opinionated.

**Comment by smyrman** [comment]
If JSONSchema conversion is successfully handled by @yanfali in #22, including deep nested schemas and without any new interfaces in the schema package itself, swagger should probably be implemented as a separate package as well in a similar style, preferable with as much reuse as possible.

Let's wait and see if #22 sees any progress.

**Comment by smyrman** [comment]
@rs. Let's reject this ticket. I think the  JSON Schema support started by @yanfali, once extended/completed, can be used to provide a JSON Schema that will be Swagger compatible, i.e. since it already contains the `readOnly` extension, and other extensions added by `Swagger`, appears less relevant.

Can create a new ticket later for generating Swagger 2.0 documentation from a `resource.Index`, if still relevant once the JSON Schema support is complete.

**Comment by rs** [comment]
Ok
